### DISCUSSION TITLE
SSR / localStorage

### DISCUSSION THREAD

**Discussion by wiesson** [firebase/firebase-js-sdk#4629] [Q&A]
I'm trying to deploy a vitejs based app (sveltekit) to vercel and I got this error during the build process:

```
> localStorage is not defined
```

I guess, to use SSR I may use firebase-admin but I was hoping there are other options :)

**Comment by akauppi** [comment]
You can tune the entry points Vite would use with the [resolve.mainFields](https://vitejs.dev/config/#resolve-mainfields) setting.

`firebase/auth` provides the following:

```
  "main": "dist/node/index.js",
  "react-native": "dist/rn/index.js",
  "browser": "dist/esm5/index.js",
  "cordova": "dist/cordova/index.esm5.js",
  "module": "dist/esm5/index.js",
  "esm2017": "dist/esm2017/index.js",
  "webworker": "dist/index.webworker.esm5.js",
```

One of them looks like `node` but is it true that we can do authentication also from command line? Haven't tried.

**Reply by wiesson** [reply]
Thanks! I'll have a look at that.

I go round in circles I tried to load auth only on client side, but then I get errors like this

```
> Firebase: Another Firebase SDK was initialized and is trying to use Auth before Auth is initialized. Please be sure to call `initializeAuth` or `getAuth` before starting any other Firebase SDK. (auth/dependent-sdk-initialized-before-auth).
```

because I'm not using a single auth reference. The compiled code simply expects the localStorage to be present:
```
... 
var BrowserLocalPersistence = function(_super) {
  __extends$1(BrowserLocalPersistence2, _super);
  function BrowserLocalPersistence2() {
    var _this = _super.call(this, localStorage, "LOCAL") || this; // <-- here
    ...
```

I have signed up for the alpha, maybe I get more info on that topic afterwards.

**Reply by akauppi** [reply]
Have seen also that error message.

Could we set up a 1-on-1 chat somewhere, to cover this? For me ([trails](https://github.com/firebase/firebase-js-sdk/discussions/4529) somewhere in these discussions...), these issues were due to the way my code was organised. As a rule of thumb, with ES Modules one should not do initialisations (`initializeAuth` or `getAuth`) in the module body. Those are run very, very, early in the import process - and it's different from how it is with `require`. Your situation may or may not be connected to this.

Good news: it's definitely solvable. :)

**Reply by wiesson** [reply]
That would be really helpful!

I'm trying to get firebase-exp working with the upcoming sveltekit (based on vite) - so server-side / client-side a bit mixed due to SSR etc. I was able to drop the localStorage error while using `initializeAuth` but I could not solve the "race-condition" 

How to reach you for a quick chat?

**Reply by akauppi** [reply]
@wiesson Drop me a line at akauppi - gmail.com and we’ll take it from there?

**Reply by thriftybabes** [reply]
was this ever resolved?
